merlin-server/constraints-dsl-compiler/src/libs/constraints-edsl-fluent-api.ts

```diff
@@ -1,87 +1,324 @@
 import * as AST from './constraints-ast.js';
 
-interface ViolationsOf extends Constraint {}
-
 export class Constraint {
   /** Internal AST node */
   public readonly __astNode: AST.Constraint;
 
-  private constructor(astNode: AST.Constraint) {
+  public constructor(astNode: AST.Constraint) {
     this.__astNode = astNode;
   }
 
-  private static new(astNode: AST.Constraint): Constraint {
-    return new Constraint(astNode);
+  public static ForbiddenActivityOverlap(activityType1: string, activityType2: string): Constraint {
+    return new Constraint({
+      kind: AST.NodeKind.ForbiddenActivityOverlap,
+      activityType1,
+      activityType2
+    })
   }
 
-  public static ViolationsOf(expression: WindowsExpression): ViolationsOf {
-    return Constraint.new({
-      kind: AST.NodeKind.ViolationsOf,
-      expression: expression.__astNode,
-    });
+  public static ForEachActivity(activityType: string, alias: string, constraint: Constraint): Constraint {
+    return new Constraint({
+      kind: AST.NodeKind.ForEachActivity,
+      activityType,
+      alias,
+      expression: constraint.__astNode
+    })
   }
 }
 
-interface True extends WindowsExpression {}
-
-export class WindowsExpression {
+export class Windows {
   /** Internal AST node */
   public readonly __astNode: AST.WindowsExpression;
 
-  private constructor(expression: AST.WindowsExpression) {
+  public constructor(expression: AST.WindowsExpression) {
     this.__astNode = expression;
   }
 
-  private static new(expression: AST.WindowsExpression): WindowsExpression {
-    return new WindowsExpression(expression);
+  public static During(alias: string): Windows {
```

Comment:
What are the aliases in these static cases referencing? Or rather, are the aliases required to exist a priori and can we enforce that in the type system?

Reply:
The aliases are created by `ForEachActivity`. example:

```typescript
export default () => {
    return Constraint.ForEachActivity(
        "activity type",
        "activity alias",
        Windows.During("activity alias").not()
     )
}
```

I don't think it can be enforced.

Reply:
I think we actually can. I think the necessary parts are here https://www.typescriptlang.org/play?#code/LAKALgngDgpgBAZTAJwJYDsDmAZVYbICGANgDwAqAfHALxzlwwAe+6AJgM5wcoaZwB+bryyMWMdlwZD0MAG4E4ALnrK4shcgDcoXSABmAV3QBjMKgD26bqgC2UYjACiTQvccUxrTsLRYANHAA4l4SPkh+OHgEJBSUlAAU+qjIPCoRfLj4RGRUgRwwJlZsKkEAlCpyFqhscADeAL56HHYOzq7uMAkA5ABMAMzdgX2DZaAtnS5ubT0DQ3Dd-QAs3WMgoCbEhBxcAEIWTJ7M3lw8kdR1oHBwAPo3AEZE7CoMdMZsMMmytdtwhOgQHQga5FdBnQxmCzIBJyEiGGAvMr1JogFGgIymcxWOD3A4wNhTTqkK6qY5hU4iTD+EkhMmSRCUrIxXKUUCJEnJVJgdKM6I5OLU4HcQrFFT7Q5BagAH3Uhls9wIgoqcCqNWRelxTHxhJmI3msgA7nBxbNRmtNdqOrq5oFDcaDrNVmUgA

Reply:
**y i k e s**

Ok I'll try that out. That'll be part of the upcoming typechecking PR. Thanks!